### PR TITLE
Don't add fixture finalizer if the value is cached

### LINKED ISSUE #1489: fixture finalizer dependency incorrect when using autouse or getfuncargvalue
Calling fixture finalizers in the correct order works by dependent fixtures registering their finalizer in their dependency (see python.FixtureDef.execute+10).
However, this only happens for fixtures explicitly requested in args, not for autouse or getfuncargvalue fixtures.
This code:

``` python
import pytest

@pytest.fixture(scope="session")
def global_dict():
    return {}

@pytest.fixture(scope="session", autouse=True)
def add_1(global_dict):
    global_dict[1] = "a"
    print("add_1 setup", global_dict)
    yield global_dict
    print("add_1 teardown", global_dict)
    del global_dict[1]

@pytest.fixture(scope="session")
def change_1_to_2(global_dict):
    global_dict[2] = global_dict.pop(1)
    print("change_1_to_2 setup", global_dict)
    yield global_dict
    print("change_1_to_2 teardown", global_dict)
    global_dict[1] = global_dict.pop(2)

def test1(change_1_to_2):
    assert change_1_to_2[2] == "a"

def test2(global_dict):
    assert global_dict[2] == "a"
```

gives with `py.test --capture=no test.py`:

```
============================= test session starts ==============================
platform linux2 -- Python 2.7.11, pytest-2.9.1, py-1.4.31, pluggy-0.3.1
rootdir: /home/amos/git/test_py, inifile: 
collected 2 items 

test.py add_1 setup {1: 'a'}
change_1_to_2 setup {2: 'a'}
..add_1 teardown {2: 'a'}
change_1_to_2 teardown {2: 'a'}
E

==================================== ERRORS ====================================
__________________________ ERROR at teardown of test2 __________________________

global_dict = {1: 'a'}

    @pytest.yield_fixture(scope="session", autouse=True)
    def add_1(global_dict):
        global_dict[1] = "a"
        print "add_1 setup", global_dict
        yield global_dict
        print "add_1 teardown", global_dict
>       del global_dict[1]
E       KeyError: 1

test.py:14: KeyError
====================== 2 passed, 1 error in 0.01 seconds =======================
```

as does this code:

``` python
import pytest

@pytest.fixture(scope="session")
def global_dict():
    return {}

@pytest.yield_fixture(scope="session")
def add_1(global_dict):
    global_dict[1] = "a"
    print "add_1 setup", global_dict
    yield global_dict
    print "add_1 teardown", global_dict
    del global_dict[1]

@pytest.yield_fixture(scope="session")
def change_1_to_2(request):
    add_1 = request.getfuncargvalue("add_1")
    add_1[2] = add_1.pop(1)
    print "change_1_to_2 setup", add_1
    yield add_1
    print "change_1_to_2 teardown", add_1
    add_1[1] = add_1.pop(2)

def test1(change_1_to_2):
    assert change_1_to_2[2] == "a"

def test2(global_dict):
    assert global_dict[2] == "a"
```

### PR DESCRIPTION
**Pull request by jakkdl**
Fixes #1489 - see https://github.com/pytest-dev/pytest/issues/1489#issuecomment-1894057766 for explanation of current behavior.

It's possible this might break test setups that (inadvertently) relies on the teardown order being as it is, but I think it's an improvement on the status quo.

The code for setup & teardown of fixtures was very hard to parse, and I added a couple comments and fixed what I think are outdated and now incorrect docstrings. But feels like a more thorough cleanup could be warranted. I have not looked into `git blame` to see if/when/how stuff got changed.

Remaining questions:
* Should finalizer be scheduled if the cached result is an exception? It previously was, I probably lean no. The change does not affect any current tests, but I could maybe write a test that asserts the behavior.
* I never managed to figure out why fixtures adds its finalizer to all fixtures it depends on, nor if this could cause setup/teardown ordering failures (I tried writing tests that would trigger it, but didn't manage to). There's definitely reasons for it though, lots of stuff breaks if I remove that code.
  * I tested moving the loop of adding the finalizer to fixtures after getting the cached value, which could plausibly remove some extraneous finalizers getting added, and that affected the result of `test_issue_519` #519 (teardown of `('test_one[arg1v2-arg2v1]', 'fix2', 'arg2v1'),` got moved one step earlier) and a similar thing in `test_parametrization_setup_teardown_ordering` (`step1-2` happening before `setup 2`), but I haven't dug into why or if/how it's bad.
  * saving the `functools.partial` call is mostly to make it more visible that it's done in several places in the function, but am down for skipping it.
* should I just make `execute` return `None`?

### CODE REVIEW

**Comment by jakkdl** [comment]
seems like the failing tests are unrelated to this PR, they're failing in several other PRs.

**Comment by bluetech** [comment]
> seems like the failing tests are unrelated to this PR, they're failing in several other PRs.

Yes, it's unrelated. If you rebase it should be good.

**Comment by nicoddemus** [comment]
> I never managed to figure out why fixtures adds its finalizer to all fixtures it depends on

IIRC, this is meant to ensure that we teardown the current fixture before its upstream fixtures, because the upstream fixture will call its finalizers and then teardown itself. 

It is a trick because fixture teardown is not structured/graph based, at least that's how I recall. For the note I also find this extremely confusing, and we have discussed in the past to replace this implicit mechanism in favor of a proper dependency graph.

**Comment by jakkdl** [comment]
> > I never managed to figure out why fixtures adds its finalizer to all fixtures it depends on
> 
> IIRC, this is meant to ensure that we teardown the current fixture before its upstream fixtures, because the upstream fixture will call its finalizers and then teardown itself.

Right. I thought that would be handled by the fact that they get torn down in reverse order of setup?
 
> It is a trick because fixture teardown is not structured/graph based, at least that's how I recall. For the note I also find this extremely confusing, and we have discussed in the past to replace this implicit mechanism in favor of a proper dependency graph.

It is at least partly graph-based, see https://github.com/pytest-dev/pytest/blob/eefc9d47fc40a9d5971a6207689c50efb1555a95/src/_pytest/runner.py#L422
But yeah the teardowns within any of session/mod/item are just flat stacks.

**Comment by RonnyPfannschmidt** [comment]
This is in part necessary to handle dynamic fixture request

**Comment by RonnyPfannschmidt** [comment]
A big part of todays complexity is the funcarg mechanism which predates fixtures and scopes

**Comment by jakkdl** [comment]
Bump :)
any blockers or other problems with the PR?

**Comment by bluetech** [comment]
@jakkdl I will try to review, but right now still mostly dealing with pytest 8 fallout.

Fair warning though, making changes to core fixture code is usually very tricky.

File: src/_pytest/fixtures.py [hunk 1]
```diff
@@ -641,11 +642,8 @@ def _compute_fixture_value(self, fixturedef: "FixtureDef[object]") -> None:
 
         # Check if a higher-level scoped fixture accesses a lower level one.
         subrequest._check_scope(argname, self._scope, scope)
-        try:
-            # Call the fixture function.
```
Comment (RonnyPfannschmidt) [inline src/_pytest/fixtures.py:645]: This looks like it originally ensured teardown even when setup failed

I suspect that we are missing a edge case tests there

**Review by RonnyPfannschmidt — COMMENTED** [review]
The overall implementation looks like an improvement

I'm slightly worried as we currently express the logic in code instead of a data structure

However at first glance the better control looks reasonable

**Review by RonnyPfannschmidt — COMMENTED** [review]
I believe we need to figure tests that include cached fixtures with failures

File: src/_pytest/fixtures.py (hunk 1 above)
Comment (RonnyPfannschmidt) [reply src/_pytest/fixtures.py:645]: The specific schedule is slightly better,

Comment (jakkdl) [reply src/_pytest/fixtures.py:645]: On further consideration, changing that seems... bad to me? It means we'll be running the finalizer multiple times for a fixture with a cached exception, even if its setup was only run once. That *is* how it worked in the past though

**Comment by jakkdl** [comment]
> I believe we need to figure tests that include cached fixtures with failures

Added a test. If I were to add the change discussed for scheduling teardown when setup failed the code for the last function would be 
```python
def test_crash_expected_setup_and_teardown() -> None:
    assert executed_crash == ["fix_crash setup", "fix_crash teardown", "fix_crash_teardown"]
```

I should also change these tests so they're robust to reordering before merging.

File: src/_pytest/fixtures.py (hunk 1 above)
Comment (RonnyPfannschmidt) [reply src/_pytest/fixtures.py:642]: Apologies, what I meant is that a single schedule is better than the current mechanism

Ideally this logic would move to a data structure instead of the code where it is

Comment (jakkdl) [reply src/_pytest/fixtures.py:642]: Right - no yeah I agree that the current implementation is quite messy and would benefit from an overhaul, but that seems out of scope for this PR.
I'm not sure I follow your original comment in that case, the `try/except` that previously was on this line is now moved one step deeper into `execute()` - and so there should be no change wrt to behaviour of scheduling finalizer if setup fails.

The only real change should :tm: be that the finalizer isn't scheduled if the value is cached (regardless of if it's an exception or not). If the setup code for the fixture fails, it's catched by the new `try/finally`.
If some other unrelated code in `execute` raises an exception the finalizer will not be scheduled, but in that case the setup has not run either.

Comment (jakkdl) [reply src/_pytest/fixtures.py:637]: So is there a problem I should address, or was your comment just a musing on how this should be generally overhauled? Or do you consider that a requirement before modifying any logic?

File: src/_pytest/fixtures.py [hunk 2]
```diff
@@ -534,6 +534,8 @@ def getfixturevalue(self, argname: str) -> Any:
         :raises pytest.FixtureLookupError:
             If the given fixture could not be found.
         """
+        # Note: This is called during setup for evaluating fixtures defined via
```
Comment (bluetech) [inline src/_pytest/fixtures.py:537]: I found the comment somewhat confusing, would remove it.

File: testing/python/test_scope_fixture_caching.py
Comment (bluetech) [inline testing/python/test_scope_fixture_caching.py:1]: We usually put such tests as pytester tests and not run them directly, this way they're isolated. Can you convert these tests to pytester tests and put them in `testing/python/fixtures.py`?

**Review by bluetech — COMMENTED** [review]
Not a real review yet, just some quick comments

File: src/_pytest/fixtures.py [hunk 3]
```diff
@@ -1071,7 +1072,15 @@ def execute(self, request: SubRequest) -> FixtureValue:
...
-        result = ihook.pytest_fixture_setup(fixturedef=self, request=request)
+        try:
+            # Setup the fixture, run the code in it, and cache the value
+            # in self.cached_result
+            result = ihook.pytest_fixture_setup(fixturedef=self, request=request)
+        finally:
+            # schedule our finalizer, even if the setup failed
+            request.node.addfinalizer(finalizer)
+
+        # note: unused
```
[the remark this answers is not in the document]
Comment (jakkdl) [reply src/_pytest/fixtures.py:1083]: Should we still keep the return value?

File: src/_pytest/fixtures.py (hunk 2 above)
Comment (jakkdl) [reply src/_pytest/fixtures.py:537]: I definitely think there should be some comment here, or have the docstring modified*, because the docstring heavily implies that this function is only used when fixture functions are evaluated dynamically. That confused me a lot when trying to understand the code.
\* I'm guessing the docstring is for the public API documentation, so whether it's used internally or not probably shouldn't be mentioned in it.

File: testing/python/test_scope_fixture_caching.py
Comment (jakkdl) [reply testing/python/test_scope_fixture_caching.py:1]: done~

File: src/_pytest/fixtures.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (bluetech) [reply src/_pytest/fixtures.py]: I'd say no, maybe some plugin is using that. But if you want to do it, can submit a separate PR, and I will check my local "plugin corpus" (a checkout of a few hundred popular plugins) if anyone is using that.

File: src/_pytest/fixtures.py (hunk 2 above)
Comment (bluetech) [reply src/_pytest/fixtures.py:534]: Yes, the docstring is intended for the public API. Maybe make the comment more explicit, e.g.

Note that in addition to the use case described in the docstring, `getfixturevalue()` is also called by pytest itself during item setup to evaluate the fixtures that are requested statically (using function parameters, autouse, etc).

Also, let's have an empty line between the comment and the code to make it clear it is not a comment for the next line but for the function.

File: testing/python/test_scope_fixture_teardown_order.py
Comment (bluetech) [inline testing/python/test_scope_fixture_teardown_order.py:1]: Can you do the same for this test as well? (Sorry I wasn't clear before)

File: src/_pytest/fixtures.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jakkdl) [reply src/_pytest/fixtures.py]: I don't think I care enough, so I'll let it be :+1:

File: testing/python/test_scope_fixture_teardown_order.py
Comment (jakkdl) [reply testing/python/test_scope_fixture_teardown_order.py:1]: oops, should've noticed it myself as well. Forgot I'd created a second test file.

File: testing/python/fixtures.py [hunk 4]
```diff
@@ -4559,3 +4559,134 @@ def test_deduplicate_names() -> None:
...
+        """
+    )
+    result = pytester.runpytest()
+    assert result.ret == 0
+
+
+def test_scope_fixture_caching_1(pytester: Pytester) -> None:
+    """
+    Make sure setup and finalization is only run once when using fixture
+    multiple times. This might be a duplicate of another test."""
```
Comment (bluetech) [inline testing/python/fixtures.py:4621]:
```suggestion
    """Make sure setup and finalization is only run once when using a fixture
    multiple times."""
```

File: src/_pytest/fixtures.py [hunk 5]
```diff
@@ -531,6 +531,11 @@ def getfixturevalue(self, argname: str) -> Any:
         :raises pytest.FixtureLookupError:
             If the given fixture could not be found.
         """
+        # Note that in addition to the use case described in the docstring,
+        # getfixturevalue() is also called by pytest itself during item setup to
```
Comment (bluetech) [inline src/_pytest/fixtures.py:535]: It's now also used in `pytest_fixture_setup` :)

**Review by bluetech — COMMENTED** [review]
Thanks a lot for the PR @jakkdl, if we can get it merged it would be a very nice simplification.

So here is my understanding:

- In issue #1489 we're seeing bad fixture teardown ordering.

- You've analyzed the situation and concluded that we're adding finalizers too eagerly, particularly what's causing the issue is that we're adding a finalizer even when the fixturedef value is already cached.

- Your PR makes it so that we don't add a finalizer when the fixture is cached.

- In the process you've also removed code that was added as a fix for #1895 (the `SubRequest._schedule_finalizers` code), because it is seemingly no longer needed (indeed, the regression test `test_getfixturevalue_teardown` passes).

So for me as a reviewer there are two questions:

- Is the addfinalizer in the cached case needed for correctness?
- Is the removal of the #1895 code OK?

### Is the removal of the #1895 code OK?

First I'm looking at the 2nd question since it's easier to examine. Here is the regression test for #1895:

```py
import pytest

@pytest.fixture(scope='session')
def resource():
    r = ['value']
    yield r
    r.pop()

@pytest.fixture(scope='session')
def inner(request):
    resource = request.getfixturevalue('resource')
    assert resource == ['value']
    yield
    assert resource == ['value']

def test_inner(inner):
    pass

def test_func(resource):
    pass
```

I added some prints in pytest to see what's going on. Here is the output *before* this PR:

```py
x.py::test_inner PASSED
TEARDOWN <Function test_inner>

x.py::test_func PASSED
TEARDOWN <Function test_func>

TEARDOWN <Module x.py>

TEARDOWN <Dir pytest>

TEARDOWN <Session  exitstatus=<ExitCode.OK: 0> testsfailed=0 testscollected=2>

FIXTURE FINISH <FixtureDef argname='resource' scope='session' baseid='x.py'> <SubRequest 'resource' for <Function test_func>>

FIXTURE FINISH <FixtureDef argname='inner' scope='session' baseid='x.py'> <SubRequest 'inner' for <Function test_inner>>

FIXTURE FINISH <FixtureDef argname='inner' scope='session' baseid='x.py'> <SubRequest 'inner' for <Function test_inner>>

FIXTURE FINISH <FixtureDef argname='resource' scope='session' baseid='x.py'> <SubRequest 'resource' for <Function test_inner>>
```

After the PR, the output is (diff):

```diff
@@ -19,10 +19,6 @@
 
 TEARDOWN <Session  exitstatus=<ExitCode.OK: 0> testsfailed=0 testscollected=2>
 
-FIXTURE FINISH <FixtureDef argname='resource' scope='session' baseid='x.py'> <SubRequest 'resource' for <Function test_func>>
-
-FIXTURE FINISH <FixtureDef argname='inner' scope='session' baseid='x.py'> <SubRequest 'inner' for <Function test_inner>>
-
 FIXTURE FINISH <FixtureDef argname='inner' scope='session' baseid='x.py'> <SubRequest 'inner' for <Function test_inner>>
 
 FIXTURE FINISH <FixtureDef argname='resource' scope='session' baseid='x.py'> <SubRequest 'resource' for <Function test_inner>>
```

The optimistic interpretation of this is that before #1895 fix, we've been adding a useless finalizer in the cached case, and the #1895 fix fixed this by adding *another*  useless finalizer to beat the previous useless finalizer. And this PR fixes it by not adding the useless finalizer in the first place.

The less optimistic interpretation is that the "useless" finalizer is not useless and is needed for correctness in some odd case, and then this PR causes some regression.

So really the second question reduces down to the first question.

### Is the addfinalizer in the cached case needed for correctness?

Unfortunately I'm out of time for today to investigate this question, so this ends with a cliffhanger...

File: testing/python/fixtures.py [hunk 6]
```diff
@@ -4559,3 +4559,134 @@ def test_deduplicate_names() -> None:
...
+            # torn down before fixture_2 - violating the rule that teardowns should
+            # happen in reverse order of setup.
+            pass
+        """
+    )
+    result = pytester.runpytest()
+    assert result.ret == 0
+
+
+def test_scope_fixture_caching_1(pytester: Pytester) -> None:
```
[the remark this answers is not in the document]
Comment (jakkdl) [reply testing/python/fixtures.py:4618]: Seems reasonable, see #12121

File: testing/python/fixtures.py [hunk 7]
```diff
@@ -4559,3 +4559,134 @@ def test_deduplicate_names() -> None:
...
+
+
+        last_executed = ""
+
+
+        @pytest.fixture(scope="module")
+        def fixture_1() -> Generator[None, None, None]:
+            global last_executed
+            assert last_executed == ""
+            last_executed = "autouse_setup"
```
[the remark this answers is not in the document]
Comment (jakkdl) [reply testing/python/fixtures.py:4584]: good catch, yep. It's only after I worked on this PR for a while I figured out that `autouse` wasn't required to surface the issue.

File: testing/python/fixtures.py [hunk 8]
```diff
@@ -4559,3 +4559,134 @@ def test_deduplicate_names() -> None:
...
+
+def test_scoped_fixture_teardown_order(pytester: Pytester) -> None:
+    """
+    Make sure teardowns happen in reverse order of setup with scoped fixtures, when
+    a later test only depends on a subset of scoped fixtures.
+    Regression test for https://github.com/pytest-dev/pytest/issues/1489
+    """
+    pytester.makepyfile(
+        """
+        from typing import Generator
```
[the remark this answers is not in the document]
Comment (jakkdl) [reply testing/python/fixtures.py:4572]: Imagine if they were though :star_struck: 
Idk, they could maybe help understanding the code, or if somebody copies the code out from the pytester-string to "real" code. I don't see much/any harm in leaving them anyhow

File: src/_pytest/fixtures.py [hunk 9]
```diff
@@ -567,9 +572,8 @@ def _compute_fixture_value(self, fixturedef: "FixtureDef[object]") -> None:
         """Create a SubRequest based on "self" and call the execute method
         of the given FixtureDef object.
 
-        This will force the FixtureDef object to throw away any previous
-        results and compute a new fixture value, which will be stored into
-        the FixtureDef object itself.
+        If the FixtureDef has cached the result it will do nothing, otherwise it will
```
[the remark this answers is not in the document]
Comment (jakkdl) [reply src/_pytest/fixtures.py:575]: oh, registering finalizers regardless of if the value is cached seems dumb... if it's cached then we've already registered a finalizer when we computed the value. And this can also cause bad teardown ordering:

```python
import pytest


@pytest.fixture(scope="module")
def fixture_1(request):
    ...

@pytest.fixture(scope="module")
def fixture_2(fixture_1):
    print("setup 2")
    yield
    print("teardown 2")

@pytest.fixture(scope="module")
def fixture_3(fixture_1):
    print("setup 3")
    yield
    print("teardown 3")

def test_1(fixture_2):
    ...
def test_2(fixture_3):
    ...

# this will reschedule fixture_2's finalizer in the parent fixture, causing it to be
# torn down before fixture 3
def test_3(fixture_2):
    ...

# trigger finalization of fixture_1, otherwise the cleanup would sequence 3&2 before 1 as normal
@pytest.mark.parametrize("fixture_1", [None], indirect=["fixture_1"])
def test_4(fixture_1):
    ...
```
this prints
```
setup 2
setup 3
teardown 2
teardown 3
```
but if we remove `test_3` we get 2-3-3-2.

But this is also a different issue+PR

**Comment by jakkdl** [comment]
> > I never managed to figure out why fixtures adds its finalizer to all fixtures it depends on,
> 
> If a fixture F1 depends on fixture F2, then F1 must be torn down before F2.
> 
> F1 guarantees this by registering its own finish as a finalizer in all fixtures it depends on. So when F2.finish() runs, it runs F1.finish() first as a finalizer.
> 
> That's a pretty brute force way to do it, but it is how it is...

Oh wait, this can cause funky ordering as well:

```py
import pytest


@pytest.fixture(scope="module", params=["a", "b"])
def fixture_1(request):
    print("setup 1 ", request.param)
    yield
    print("teardown 1", request.param)


@pytest.fixture(scope="module")
def fixture_2():
    print("setup 2")
    yield
    print("teardown 2")


@pytest.fixture(scope="module")
def fixture_3(fixture_1):
    print("setup 3")
    yield
    print("\nteardown 3")


def test_1(fixture_1, fixture_2, fixture_3): ...
```

```
setup 1  a
setup 2
setup 3
.
teardown 3
teardown 1 a
setup 1  b
setup 3
.
teardown 3
teardown 1 b
teardown 2 <-- 2 is torn down out of order
```

That's definitely a different issue though, and not sure how to tackle that - if at all realistic.

**Review by bluetech — APPROVED** [review]
After thinking about the first question mentioned above, I can't think of a reason why we would want to register the finalizer again in the cached case, so I'm going to merge this. This doesn't mean it's entirely safe, I still have a bit of a [Chesterton's fence](https://fs.blog/chestertons-fence/) feeling, but I guess we'll find out :)

My line of reasoning is the following (mostly for self reference...):

- If the fixture is cached, its finalizer has already been registered in the relevant node. So adding the finalizer in the cached case can't be about actually *ensuring* it is finalized, it's only about the *order* of finalization.

- The desired finalization order is (generally?) reversed setup order. The setup order is determined by the uncached case, so the effect of adding the finalizer in the cached case can basically only be to get the finalizer to run *earlier* than the reversed setup order (or have no effect).

- So the question is, is there nonetheless some reason why we'd want to diverge from reversed setup order? Potential suspects are: higher scope, parametrizations, dynamic `getfixturevalue`. Here things become a bit blurry for me, but let's see:

- Dynamic `getfixturevalue` - I don't think the static/dynamic distinction is relevant for this, since this is only about the "dynamic" part of the fixture mechanism.

- Parametrizations - so parametrization basically "kills" a vertex in the fixture DAG and then all dependent vertexes need to be invalidated, while the rest of the graph remains. There are three cases:

    - The fixture itself is parametrized -- in this case the cache key is different so this case should not be relevant.

    - The fixture depends on a parametrized fixture - for this we have the loop at the beginning of `execute` so shouldn't matter.

   - A parametrized fixture depends on the fixture - don't think it should matter to the fixture.

- Higher scopes - I don't think it should affect reversed setup order.

File: src/_pytest/fixtures.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (Zac-HD) [reply src/_pytest/fixtures.py:595]: @jakkdl would you mind opening a fresh issue for this case and the one you describe in https://github.com/pytest-dev/pytest/pull/11833#issuecomment-1999818465?

[the remark this answers is not in the document]
Comment (jakkdl) [reply src/_pytest/fixtures.py:595]: Done: #12134 #12135

**Comment by asottile-sentry** [comment]
haven't figured out what about it yet -- but something about this PR makes sentry's testsuite very flaky

annoyingly it's not very reproducible beyond running a whole suite a bunch of times so my "bisect" has taken [a few days](https://github.com/getsentry/sentry/pull/76935) (I sorta manually bisected against `git log --oneline --first-parent 8.1.0..8.2.0 -- src`)

the failure mode looks like https://github.com/getsentry/sentry/actions/runs/10905480202/job/30267377794?pr=76935 -- the vast majority of the exceptions are in a per-test teardown 

don't suppose anyone has more specific ideas about how I should narrow down what's causing this? I tried looking at the `--setup-plan` and the only differences in that seems to be in session-scoped fixtures teardown (but that's long long after the errors start happening)

___

my current hypothesis (with absolutely no evidence because I haven't found a consistent reproduction) is the first test failure (which happens during teardown) somehow interrupts the rest of the teardowns meaning the database connections never get cleaned up (and then subsequent tests which _check_ that error).  the test itself is flaky but mostly paved over with retries -- but that shouldn't bring down the rest of the suite with it!

<details>
<summary>first teardown failure, kinda long</summary>

```
_ ERROR at teardown of GroupDetailsTest.test_plugin_external_issue_annotation __
src/sentry/db/postgres/decorators.py:91: in inner
    return func(self, sql, *args, **kwargs)
src/sentry/db/postgres/base.py:85: in execute
    return self.cursor.execute(sql)
E   psycopg2.errors.ForeignKeyViolation: insert or update on table "sentry_projectcounter" violates foreign key constraint "sentry_projectcounter_project_id_90383de8_fk_sentry_project_id"
E   DETAIL:  Key (project_id)=(4554717002661889) is not present in table "sentry_project".

During handling of the above exception, another exception occurred:
.venv/lib/python3.12/site-packages/django/db/backends/utils.py:103: in _execute
    return self.cursor.execute(sql)
src/sentry/db/postgres/decorators.py:77: in inner
    raise_the_exception(self.db, e)
src/sentry/db/postgres/decorators.py:75: in inner
    return func(self, *args, **kwargs)
src/sentry/db/postgres/decorators.py:18: in inner
    return func(self, *args, **kwargs)
src/sentry/db/postgres/decorators.py:93: in inner
    raise type(e)(f"{e!r}\nSQL: {sql}").with_traceback(e.__traceback__)
src/sentry/db/postgres/decorators.py:91: in inner
    return func(self, sql, *args, **kwargs)
src/sentry/db/postgres/base.py:85: in execute
    return self.cursor.execute(sql)
E   psycopg2.errors.ForeignKeyViolation: ForeignKeyViolation('insert or update on table "sentry_projectcounter" violates foreign key constraint "sentry_projectcounter_project_id_90383de8_fk_sentry_project_id"\nDETAIL:  Key (project_id)=(4554717002661889) is not present in table "sentry_project".\n')
E   SQL: SET CONSTRAINTS ALL IMMEDIATE

The above exception was the direct cause of the following exception:
.venv/lib/python3.12/site-packages/django/test/testcases.py:372: in _setup_and_call
    self._post_teardown()
src/sentry/testutils/cases.py:415: in _post_teardown
    super()._post_teardown()
.venv/lib/python3.12/site-packages/django/test/testcases.py:1202: in _post_teardown
    self._fixture_teardown()
.venv/lib/python3.12/site-packages/django/test/testcases.py:1455: in _fixture_teardown
    connections[db_name].check_constraints()
.venv/lib/python3.12/site-packages/django/db/backends/postgresql/base.py:482: in check_constraints
    cursor.execute("SET CONSTRAINTS ALL IMMEDIATE")
.venv/lib/python3.12/site-packages/django/db/backends/utils.py:122: in execute
    return super().execute(sql, params)
.venv/lib/python3.12/site-packages/sentry_sdk/utils.py:1718: in runner
    return original_function(*args, **kwargs)
.venv/lib/python3.12/site-packages/django/db/backends/utils.py:79: in execute
    return self._execute_with_wrappers(
.venv/lib/python3.12/site-packages/django/db/backends/utils.py:92: in _execute_with_wrappers
    return executor(sql, params, many, context)
src/sentry/testutils/hybrid_cloud.py:130: in __call__
    return execute(*params)
.venv/lib/python3.12/site-packages/django/db/backends/utils.py:100: in _execute
    with self.db.wrap_database_errors:
.venv/lib/python3.12/site-packages/django/db/utils.py:91: in __exit__
    raise dj_exc_value.with_traceback(traceback) from exc_value
.venv/lib/python3.12/site-packages/django/db/backends/utils.py:103: in _execute
    return self.cursor.execute(sql)
src/sentry/db/postgres/decorators.py:77: in inner
    raise_the_exception(self.db, e)
src/sentry/db/postgres/decorators.py:75: in inner
    return func(self, *args, **kwargs)
src/sentry/db/postgres/decorators.py:18: in inner
    return func(self, *args, **kwargs)
src/sentry/db/postgres/decorators.py:93: in inner
    raise type(e)(f"{e!r}\nSQL: {sql}").with_traceback(e.__traceback__)
src/sentry/db/postgres/decorators.py:91: in inner
    return func(self, sql, *args, **kwargs)
src/sentry/db/postgres/base.py:85: in execute
    return self.cursor.execute(sql)
E   django.db.utils.IntegrityError: ForeignKeyViolation('insert or update on table "sentry_projectcounter" violates foreign key constraint "sentry_projectcounter_project_id_90383de8_fk_sentry_project_id"\nDETAIL:  Key (project_id)=(4554717002661889) is not present in table "sentry_project".\n')
E   SQL: SET CONSTRAINTS ALL IMMEDIATE
```

</details>

**Comment by bluetech** [comment]
@asottile-sentry What do you use for retries?

**Comment by jakkdl** [comment]
@asottile-sentry as I said at the start of the PR:
> It's possible this might break test setups that (inadvertently) relies on the teardown order being as it is, but I think it's an improvement on the status quo.

To debug I would suggest logging extensively the order that your fixtures are set up and torn down (before and after this PR) and also watch out for #12134. It's possible that you need to add inter-fixture dependencies to ensure that they're handled correctly.

**Comment by asottile-sentry** [comment]
> @asottile-sentry What do you use for retries?

`pytest-rerunfailures` -- though I wish we didn't because paving over retries just doesn't seem like a scalable solution if we ever want to _reduce_ flakiness 😆 

> @asottile-sentry as I said at the start of the PR:
> 
> > It's possible this might break test setups that (inadvertently) relies on the teardown order being as it is, but I think it's an improvement on the status quo.
> 
> To debug I would suggest logging extensively the order that your fixtures are set up and torn down (before and after this PR) and also watch out for #12134. It's possible that you need to add inter-fixture dependencies to ensure that they're handled correctly.

are you saying that `--setup-plan` doesn't tell me how the fixtures will actually order? (there were minor changes in that order as I noted above -- but far far after the failures themselves)

**Comment by jakkdl** [comment]
Oh, knowing about `--setup-plan` would've been better than liberal use of `print` to debug this and related issues in the first place :upside_down_face: 

This PR really shouldn't change anything other than teardown ordering in some cases, I have no clue why it would impact anything else. But the fixture system is quite messy so ???

**Comment by bluetech** [comment]
> pytest-rerunfailures

I suspected this -- rerunfailures is great but it adds another layer of complexity to such issues, especially involving teardown failures, because it does substantial hacking around the pytest runner internals. If you're able to try, do the cascading failures still happen without rerunfailures? Knowing this will tell us where to look.

> though I wish we didn't because paving over retries just doesn't seem like a scalable solution if we ever want to reduce flakiness 😆

Indeed...

Also, teardown errors should be avoided whenever possible, semantically they're a mess to deal with.

**Comment by asottile-sentry** [comment]
> > pytest-rerunfailures
> 
> I suspected this -- rerunfailures is great but it adds another layer of complexity to such issues, especially involving teardown failures, because it does substantial hacking around the pytest runner internals. If you're able to try, do the cascading failures still happen without rerunfailures? Knowing this will tell us where to look.
> 
> > though I wish we didn't because paving over retries just doesn't seem like a scalable solution if we ever want to reduce flakiness 😆
> 
> Indeed...
> 
> Also, teardown errors should be avoided whenever possible, semantically they're a mess to deal with.

turning off rerunfailures does seem to fix this -- we're also on an old version of that so I'm going to try upgrading it too

**Comment by asottile-sentry** [comment]
> > > pytest-rerunfailures
> > 
> > 
> > I suspected this -- rerunfailures is great but it adds another layer of complexity to such issues, especially involving teardown failures, because it does substantial hacking around the pytest runner internals. If you're able to try, do the cascading failures still happen without rerunfailures? Knowing this will tell us where to look.
> > > though I wish we didn't because paving over retries just doesn't seem like a scalable solution if we ever want to reduce flakiness 😆
> > 
> > 
> > Indeed...
> > Also, teardown errors should be avoided whenever possible, semantically they're a mess to deal with.
> 
> turning off rerunfailures does seem to fix this -- we're also on an old version of that so I'm going to try upgrading it too

upgrading it did not fix the problem -- so I believe there's something amiss with rerunfailures and pytest>=8.2 -- potentially with (partial?) missing teardowns

**Comment by asottile-sentry** [comment]
I think I'm getting closer -- I now have a local reproduction and it seems to need:
- a flaky test
- pytest-rerunfailures (with reruns)
- pytest-sentry (with sentry configured)

hopefully I'll be able to find the problem from that!

**Comment by asottile-sentry** [comment]
finally narrowed it down to a minimal example -- and reconfirmed that this is the patch which regresses it via bisect

```python
from unittest import TestCase

g = True


class TestCaseTest(TestCase):
    @classmethod
    def tearDownClass(cls):
        print('class teardown!')

    def test(self):
        global g
        print('test!')
        if g:
            print('flaky fail!')
            g = False
            raise AssertionError('#')
```

this is what it should produce:

```console
$ ./venv/bin/pytest --reruns=2 -s t.py
============================ test session starts ============================
platform darwin -- Python 3.12.2, pytest-8.1.2, pluggy-1.5.0
rootdir: /private/tmp/y
plugins: rerunfailures-14.0
collected 1 item                                                            

t.py test!
flaky fail!
Rtest!
class teardown!
.

======================== 1 passed, 1 rerun in 0.01s =========================
```

and this is what it produces after this patch (note that `class teardown!` is missing):

```console
$ ./venvnew/bin/pytest --reruns=2 -s t.py
============================ test session starts ============================
platform darwin -- Python 3.12.2, pytest-8.2.0, pluggy-1.5.0
rootdir: /private/tmp/y
plugins: rerunfailures-14.0
collected 1 item                                                            

t.py test!
flaky fail!
Rtest!
.

======================== 1 passed, 1 rerun in 0.01s =========================
```

**Comment by jakkdl** [comment]
before this patch redundant copies of teardowns were added (which is what caused erronous ordering), so maybe `rerunfailures` ends up removing some teardowns and was relying on there being duplicate copies in some way?

**Comment by jakkdl** [comment]
https://github.com/pytest-dev/pytest-rerunfailures?tab=readme-ov-file#compatibility

pytest-rerunfailures does not official support class-scoped fixtures, which unittest `tearDownClass` might count as? Although some recent issues suggest that might have been resolved.

I think you may be another instance of https://github.com/pytest-dev/pytest-rerunfailures/issues/267 though, this PR was released as part of 8.2
